### ISSUE TITLE
Handlebars helpers: Stack Config file path not available

### ISSUE THREAD

**Issue by mfonsen** [takomo-io/takomo#76]
Handlebars helpers in Stack Group configuration can't access `this.stackGroup.path`. Would there be another way of finding out the path of the Stack Config file that is being deployed?

### Use case: 
I'm trying to migrate from a Sceptre 2.x project that relied on the idea of cascading configs and reading parameters from file path. The path structure in this project is /{applicationName}/{region}/{stage}/{stack base name}/{stack base name}.yaml. 

In `config.yaml`s on various levels I was able to fetch for example stage by using the following Ninja template `{% set p = stack_group_path.split('/') %}{{ p[1] }}`. The path would be different for each deployed stack.

### Takomo version
1.3.1. The project lives in a monorepo that is stuck to Node versions supported by AWS Lambda.

**Comment by hmeltaus** [comment]
We could expose more information about stackGroup to config files but before implementing that could you provide some snippets of your configuration so I could see how you are actually using the parameters?

I'd also like to know if you are using the values read from the files as stack input parameters or for something else. 

If the parameter values do not contain anything secret, you could move them from external files to stack configuration.

**Comment by mfonsen** [comment]
I'm pretty sure there's a better way. 

Using external files would be difficult as most of the parameters contain dynamic parts derived from the path. You could define a series of files but that would mean partial duplication.

This pattern is not perfect. In this system with Sceptre you end up getting lots of duplication at leaf level. Especially as I haven't found a way to compose a list of parameters from multiple sources. In Stack Config each parameter is "imported" individually.

I hope the following snippets would give a better idea how I'm are using Sceptre. It's in Sceptre format if you don't mind.

This can be launched with `sceptre --var suffix=foo --var ... launch --yes killerapp/dev`

It would have been a better idea to put stage and application name the other way round so you could launch multiple applications in parallel if that would be needed.

### /config.yaml
    # <% set o = stack_group_path.split(stack_group_path.split('/')[2])[1][1:] %><< o >> is used to get the fourth one. 
    # It assumes there are only four directories and that third one can be be used to split the string
    custom:
      application: '{% set p = stack_group_path.split('/') %}{{ p[0] }}'
      stage: '{% set p = stack_group_path.split('/') %}{{ p[1] }}'
      stack: '{% set o = stack_group_path.split(stack_group_path.split('/')[2])[1][1:] %}{{ o }}'
      suffix: "{{ var.suffix | default("")}}"
      stack_name: '{% set o = stack_group_path.split(stack_group_path.split('/')[2])[1][1:] %}{{ o }}-{% set p = stack_group_path.split('/') %}{{ p[2] }}-{% set p = stack_group_path.split('/') %}{{ p[1] }}{{ var.suffix | default("")}}'
      foobar_table_name:  'foobar-{% set p = stack_group_path.split('/') %}{{ p[2] }}-{% set p = stack_group_path.split('/') %}{{ p[1] }}{{ var.suffix | default("")}}-table'
      global_region: 'us-east-1'
    region: '{% set p = stack_group_path.split('/') %}{{ p[2] }}'
    project_code: '{% set p = stack_group_path.split('/') %}{{ p[0] }}'

### /killerapp/dev/config.yaml
    custom_account:
      alarm_email: 'foo@example.com'

### /killerapp/dev/foobar/foobar.yaml
    template_path: foobar.yaml
    stack_name: {{ stack_group_config.custom.stack_name }}
    parameters:
      FunctionPackageFoobar: {{ var.fp_foobar }}
      DeploymentBucketName: {{ stack_group_config.custom.deployment_bucket_name }}
      Stage: {{ stack_group_config.custom.stage }}
      Suffix: {{ stack_group_config.custom.suffix }}
      FoobarTableName:
        {{ stack_group_config.custom.foobar_table_name }}
      AlarmsSnsArn:
        !stack_output_external another-{{ stack_group_config.region }}-{{ stack_group_config.custom.stage }}::AlarmsSnsArn 
      AlarmEmail: {{ stack_group_config.custom_account.alarm_email }}

**Comment by hmeltaus** [comment]
It seems that exposing stack group properties in its config file is the way to go here. I'll implement this feature also to Takomo 1.x but I would recommend you to check if there's a way to use Takomo 2.x in your project.

**Comment by mfonsen** [comment]
Thank you. 

We'll definitely look into Node version switching. But probably AWS Lambda will gain Node 14 support before that.

In the mean time I figured out that I could use Handlebars partials to get around this problem. One file for parameters and one for top level properties. This seems doable and might be even easier than the original Sceptre solution. Unfortunately unlike Stack Group configs this solution does not support cascading configs pattern.

In general it would be helpful if there would be an opinionated comprehensive configuration. Although the documentation is very good and contains many bits of examples, it's not easy to see what would be a good way of using Takomo.

**Comment by hmeltaus** [comment]
Now, in stack group configuration files (all config.yml files within the stacks dir) you can use the following properties:

- stackGroup.name - Name of the stack group
- stackGroup.path - Path of the stack group
- stackGroup.pathSegments - Path of the stack group split into an array using `/` as a seperator.

For example, if you have a stack group file `stacks/killerapp/dev/config.yml`, then the above properties would have the following values:

- stackGroup.name = `dev`
- stackGroup.path = `/killerapp/dev`
- stackGroup.pathSegments = `[ killerapp, dev ]`

You might need to implement some custom helpers if you need to use these values in more complex way.

This feature was released in Takomo versions `1.4.0` and `2.6.0`.

**Comment by mfonsen** [comment]
I finally got into testing this with `1.4.1`. In `/config.yml` I don't have the path of the stack configuration file available. 

One of Sceptre's great features is cascading config. All configuration gets merged together before it's rendered. Or so it feels like. Takomo seems to render before merging. 

### Test files:
- /config.yml - data.topPath: topPath{{this.stackGroup.path}}
- /killerapp/dev/eu-west-1/foo/config.yml - data.middlePath: middlePath{{this.stackGroup.path}}
- /killerapp/dev/eu-west-1/foo/vpc.yml - data.bottomPath: bottomPath{{this.stackGroup.path}}
- /templates/vpc.hbs - ({{log this}})

### Results:
```
  stack: {
    path: '/killerapp/dev/eu-west-1/foo/vpc.yml/eu-west-1',
    template: 'vpc.hbs',
    region: 'eu-west-1', ...
    data: {
      topPath: 'topPath/',
      middlePath: 'middlePath/killerapp/dev/eu-west-1/foo',
      bottomPath: 'bottomPath/killerapp/dev/eu-west-1/foo'
    }
```

Did I do something wrong?
  
Would there be a Takomo compatible way of achieving the same results? 

For example with the following simplified use case: In /config.yml I would like to define once CORS address for all APIs. The address is in format www.{region}.{stage}.{application}.example.com. In the test setup below all of these parameters are available as entries in the path.

**Comment by hmeltaus** [comment]
I'll expose `stack` object to stack configuration files but there isn't much information that can be included to it because most of the stack configuration is defined in that file itself and can't be accessed before the file is actually parsed. That's why there is `stack` object available in stack template files but not in stack configuration files.

I think the new `stack` object in stack configuration should include the stack path because that is available at that point and can't be changed in the stack configuration itself. What other information could be useful?

The `stack` object currently available in stack templates is the only place where you can see the current stack region. There you won't see other regions anymore, although it is possible to expose those too if that would be help.

Your custom helper solution sounds interesting and I'd like to see how it looks like. So, if it's possible, could you share it here?

Also, I feel that our discussion here is a bit too slow paced and therefore created a room in Gitter. Feel free to send questions and ideas there, too. https://gitter.im/takomo-io/community

**Comment by mfonsen** [comment]
Makes sense.

Here's an example helper that resolves variables on stack config level:

```/**
 * Usage example in Stack Configuration (not stack Group Configuration):
 * depends:
 *  - /app3/{{ app-context "region" }}/some/some.yml
 * parameters:
 *     SesArn: 'arn:aws:ses:{{ app-context "region" }}:{{ stackGroup.accountIds }}:identity/no-reply@{{#unless (is-stage "prod")}}{{ app-context "stage" }}.{{/unless}}killerapp.com'
 */
module.exports = {
  name: "app-context",
  fn: function (propertyName) {
    const pathArray = this.stackGroup.path.split("/").slice(1)
    if (pathArray.length !== 3)
      throw new Error(`Invalid path for app-context: this.stackGroup.path`)

    const application = pathArray[0]
    const region = pathArray[1]
    const stack = pathArray[2]
    const suffix = this.var.suffix || ""
    const stage = this.var.stage
    if (stage === undefined) throw new Error("Missing var.stage")

    const data = {
      application,
      region,
      stage,
      stack,
      suffix,
      stackName: `${stack}-${region}-${stage}${suffix}`,
      deploymentBucket: `app-${region}`,
      tableName: `app-${region}-${stage}${suffix}-table`
    }
    if (data[propertyName] === undefined)
      throw new Error(`app-context: invalid property name: ${propertyName}`)
    return data[propertyName]
  }
}
```

Compared to example posted earlier above, I merged all config files for different stages so stage is now a variable. If I would have access to stack level `region` I would be able to do the same for region folders as well. That I can't do with this helper as it can't access stack object.

Regarding what information would be nice to have: the name of the stack configuration file name. In the current setup my paths are applicationName/regionName/stackName/stackName.yml. StackGroup object now exposes path and pathSegments but not the name of the file. The name of the file without extension would be nice to have.

Sorry it took too long to reply. I was able to take first application in to production with Takomo. Wen't well and learned many new things :) Hopefully I get to document something to issues or to the Gitter.
